sdks/java/harness/jmh/src/main/java/org/apache/beam/fn/harness/jmh/ProcessBundleBenchmark.java

```diff
@@ -283,4 +300,207 @@ public void testLargeBundle(TrivialTransform trivialTransform) throws Exception
     }
     assertEquals(3_000, outputValuesCount.getAndSet(0));
   }
+
+  @State(Scope.Benchmark)
+  public static class StatefulTransform extends SdkHarness {
+    final BundleProcessor processor;
+    final ExecutableProcessBundleDescriptor descriptor;
+    final StateRequestHandler nonCachingStateRequestHandler;
+    final StateRequestHandler cachingStateRequestHandler;
+
+    @SuppressWarnings({
+      "unused" // TODO(BEAM-13271): Remove when new version of errorprone is released (2.11.0)
+    })
+    private static class StatefulOutputZeroOneTwo
+        extends DoFn<KV<String, String>, KV<String, String>> {
+      private static final String STATE_ID = "bagState";
+
+      @StateId(STATE_ID)
+      private final StateSpec<BagState<String>> bagStateSpec = StateSpecs.bag(StringUtf8Coder.of());
+
+      @ProcessElement
+      public void process(ProcessContext ctxt, @StateId(STATE_ID) BagState<String> state) {
+        int size = Iterables.size(state.read());
+        if (size == 3) {
+          for (String value : state.read()) {
+            ctxt.output(KV.of(ctxt.element().getKey(), value));
+          }
+          state.clear();
+        } else {
+          state.add(ctxt.element().getValue());
+        }
+      }
+    }
+
+    private static class ToKeyAndValueDoFn extends DoFn<byte[], KV<String, String>> {
+      @ProcessElement
+      public void process(ProcessContext ctxt) {
+        ctxt.output(KV.of("key", "value"));
+      }
+    }
+
+    public StatefulTransform() {
+      try {
+        Pipeline p = Pipeline.create();
+        p.apply("impulse", Impulse.create())
+            .apply("toKeyAndValue", ParDo.of(new ToKeyAndValueDoFn()))
+            .apply("stateful", ParDo.of(new StatefulOutputZeroOneTwo()))
+            // Force the output to be materialized
+            .apply("gbk", GroupByKey.create());
+
+        RunnerApi.Pipeline pipelineProto = PipelineTranslation.toProto(p);
+        FusedPipeline fused = GreedyPipelineFuser.fuse(pipelineProto);
+        checkState(fused.getFusedStages().size() == 2, "Expected exactly two fused stages");
```

Comment:
What are the two expected stages? 1. impulse+toKeyAndValue 2. stateful+gbk?

My understanding is that `toKeyAndValue` is just a dummy transform that is added only for pipeline construction, and then its stage is not actually executed. Is that correct? And can we add a comment to clarify it?